runtime-modules/project-token/src/lib.rs

```diff
@@ -803,6 +813,165 @@ decl_module! {
             Self::deposit_event(RawEvent::RevenueSplitLeft(token_id, member_id, staking_info.amount));
             Ok(())
         }
+
+        /// Mint desired `token_id` amount into user account via JOY exchnage
+        /// Preconditions
+        /// - origin, member_id pair must be a valid authentication pair
+        /// - token_id must exist
+        /// - user usable JOY balance must be enough for buying (+ existential deposit)
+        /// - deadline constraint respected if provided
+        /// - slippage tolerance constraints respected if provided
+        /// - token total supply and amount value must be s.t. `eval` function doesn't overflow
+        ///
+        /// Postconditions
+        /// - `amount` CRT minted into account (which is created if necessary with existential deposit transferred to it)
+        /// - respective JOY amount transferred from user balance to amm treasury account
+        /// - event deposited
+        #[weight = 100_000_000] // TODO: adjust weight
+        fn buy_on_amm(origin, token_id: T::TokenId, member_id: T::MemberId, amount: <T as Config>::Balance, deadline: Option<<T as timestamp::Config>::Moment>, slippage_tolerance: Option<(Permill, JoyBalanceOf<T>)>) -> DispatchResult {
+            if amount.is_zero() {
+                return Ok(()); // noop
+            }
+
+            let sender = ensure_signed(origin.clone())?;
+
+            T::MemberOriginValidator::ensure_member_controller_account_origin(
+                origin,
+                member_id
+            )?;
+
+            let token_data = Self::ensure_token_exists(token_id)?;
+            let curve = token_data.amm_curve.ok_or(Error::<T>::NotInAmmState)?;
+
+            let user_account_data_exists = AccountInfoByTokenAndMember::<T>::contains_key(token_id, &member_id);
+            let amm_treasury_account = Self::amm_treasury_account(token_id);
+            let price = curve.eval::<T>(amount, token_data.total_supply, AmmOperation::Buy)?;
```

Comment:
This is an old debate.
When it came up I suggested 2 but Bedeho expresses his desire for 1. That's why it is implemented like this
I am in favour of 2 because in that case `b` has the meaning of "price for the first token bought on the amm".
This can (and should be enforced) allow to set `b=last_sale_price`

Reply:
I passed the `provided_supply` as an argument.
I will refactor the function as part of the next AMM PR, with all the constraints